controllers/user_keys_test.go

```diff
@@ -180,9 +198,80 @@ func (suite *UserKeysTestSuite) TestSaveKeyInvalidHex() {
 }
 
 func (suite *UserKeysTestSuite) TestGetKeyNotFound() {
-	req := httptest.NewRequest(http.MethodGet, "/v2/keys/nonexistent", nil)
+	req := httptest.NewRequest(http.MethodGet, "/v2/keys/accounts/nonexistent", nil)
 	req.Header.Set("Authorization", "Bearer "+suite.authToken)
 	resp := util.ExecuteTestRequest(req, suite.router)
 
 	suite.Equal(http.StatusNotFound, resp.Code)
 }
+
+func (suite *UserKeysTestSuite) createMaxKeys(service string) {
+	requestBody := controllers.UserKeyStoreRequest{
+		Service:     service,
+		KeyMaterial: "0123456789abcdef",
+	}
+
+	for i := 0; i < datastore.MaxUserKeysPerService; i++ {
+		requestBody.KeyName = fmt.Sprintf("test_key_%d", i)
+
+		req := util.CreateJSONTestRequest("/v2/keys", requestBody)
+		req.Header.Set("Authorization", "Bearer "+suite.authToken)
+		resp := util.ExecuteTestRequest(req, suite.router)
+
+		suite.Equal(http.StatusNoContent, resp.Code)
+	}
+}
+
+func (suite *UserKeysTestSuite) TestSaveKeyLimitExceeded() {
+	// First, store some keys for a different service to ensure per-service isolation
+	suite.createMaxKeys("sync")
+
+	// Now store MaxUserKeysPerService keys for the accounts service
+	suite.createMaxKeys("accounts")
+
+	// Try to store one more key for accounts service - should fail
```

Comment:
I wonder whether it makes sense to have a limit for Accounts itself. It doesn't really give us the same benefits as with the dependent services since we fully control Accounts.

Maybe we only apply the limit to the other services?

Reply:
Correct me if I'm wrong, but I think the main benefit of the limit is to avoid abuse scenarios. Thoughts?

Reply:
Right, "abuse" in the sense of dependent services using this in a way it was not intended to. I'm not worried about Accounts using it incorrectly because ultimately Accounts is the one that defines how it's used :)

Reply:
discussed via DM